### PR TITLE
Javap for Java 7 (Fixes SI-4936)

### PR DESCRIPTION
**Pull request by som-snytt**
Add support for reflective invocation of javap under jdk7,
using the quasi-javax.tools API.

Under -Ygen-javap, warn if you can't.

Since JAVA_HOME is used to locate tools.jar, the script is
updated to convert it for cygwin.

A par-test awaits a fix for SI-6026 and perhaps consensus on usage of tests requiring tools.jar, e.g., whether a missing tools.jar should fail the test or skip it.  Jenkins knows if java.home is set correctly.  That's why he's Jenkins.

But this will be moot after the asm textifier.

### CODE REVIEW

File: src/compiler/scala/tools/util/JavaxTools.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  val collector = new DiagnosticCollector[JavaFileObject]
+  override def report(diagnostic: Diagnostic[_ <: JavaFileObject]) { collector report diagnostic }
+  /** All diagnostics in the collector. */
+  def diagnostics: Iterable[Diagnostic[_ <: JavaFileObject]] = collector.getDiagnostics.asScala
+  /** Locale for diagnostic messages, null by default. */
+  def locale: Locale = null
+  /** All diagnostic messages. */
+  def messages = (diagnostics map (_ getMessage locale)).toList
+  /** Count the errors. */
+  def errorCount: Int = (0 /: diagnostics)((sum, d) => sum + (if (d.getKind == ERROR) 1 else 0))
```
Comment (retronym) [inline src/compiler/scala/tools/util/JavaxTools.scala:77]: You can have someone else fold this laundry:

`diagnostics count (_.getKind == ERROR)`

**Comment by paulp** [comment]
Thanks a ton for undertaking this. I'd like to reconcile these changes with a pull request of my own, #1573, where I have named you the reviewer if you don't mind. The set of plausible repl reviewers is small.

**Comment by som-snytt** [comment]
OK.  I started looking at why javap Bippy stopped working, implicitly determined to learn how repl works, and got as far as javap $line3.$read$$iw$$iw$$Bippy or whatever. It's one of those hikes where the summit looks like it's right over there but never gets any closer.

File: src/compiler/scala/tools/util/JavaxTools.scala (hunk 1 above)
Comment (som-snytt) [inline src/compiler/scala/tools/util/JavaxTools.scala:77]: Thanks. I inherited that code from six months ago, when I didn't know how to use scaladoc.  For the record, folding laundry can be quite meditative.

**Comment by paulp** [comment]
Well, it took me years to figure it all out well enough to write #1573. But you might find that brings the summit closer.

**Comment by paulp** [comment]
Oh, I forgot to mention that my recent commit of -Yrepl-outdir was motivated in large part by a desire to take the easy route out. With the classfiles exposed in the filesystem, we could just run javap.

File: src/compiler/scala/tools/util/JavaxTools.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/* NSC -- new Scala compiler
+ * Copyright 2005-2012 LAMP/EPFL
+ * @author Paul Phillips
```
Comment (paulp) [inline src/compiler/scala/tools/util/JavaxTools.scala:3]: Hey, get a name of your own.

Perhaps more is going on here than I see at a glance, but I think you can lose almost the entire file by way of

```
import javax.tools._
class FM extends ForwardingJavaFileManager[StandardJavaFileManager](ToolProvider.getSystemJavaCompiler.getStandardFileManager(null, null, null))
```

Of course I had to find and report https://issues.scala-lang.org/browse/SI-6617 to write that line.

Comment (som-snytt) [inline src/compiler/scala/tools/util/JavaxTools.scala:3]: That's merely in the sense of "the author of our woe." "For there again, against the glass, as if to blight his confession and stay his answer, was the hideous author of our woe--the white face of damnation."  That's the first thing I've ever felt like "re-tweeting."  Or, if you prefer, "the author of our faith."  It amounts to the same.
Re: filemanager, that's an artifact of not wanting to forward calls to fail silently; I want to know what protocol the tool demands. Footnote, javap has its own FileManager (minor extension).  The unsupported API can change, as well.  At least I removed all my printlns.  I'll test the FwdingJFM.

**Comment by som-snytt** [comment]
I rebased and lost the extra file.  It seems to work except that the output is reduced to a single clipped line.  Exposing and using new JavapClass(addToolsJarToLoader(), intp.out) works fine.  Before I look, can you tell me what changed?  Sorry to sound lazy.  I hate every minute of my life I've lost to looking inside PrintWriter.

**Comment by paulp** [comment]
I don't know offhand, but if you push the rebase here or somewhere I can take a look.

**Comment by som-snytt** [comment]
I didn't want to make you look.  Except the way the kids say, "Made you look!"

**Comment by som-snytt** [comment]
The change to "remove unused imports" makes it non-fast-forward, so closing in favor of:
https://github.com/scala/scala/pull/1637
which includes everyone's fine suggestions, hints and offers I couldn't refuse.

**Comment by lrytz** [comment]
it's ok to `push -f` to your own pull request branch, si jamais.

**Pull request closed without merging**